ams/server/src/main/java/com/netease/arctic/server/metrics/MetricRegistry.java

```diff
@@ -0,0 +1,133 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Consumer;
+
+/** A registry of amoro metric. */
+public class MetricRegistry implements MetricSet {
+
+  private final List<MetricRegisterListener> listeners = new CopyOnWriteArrayList<>();
+
+  private final ConcurrentMap<MetricKey, Metric> registeredMetrics = Maps.newConcurrentMap();
+  private final Map<String, Pair<MetricDefine, Integer>> definedMetric = Maps.newConcurrentMap();
+
+  /**
+   * Add metric registry listener
+   *
+   * @param listener Metric registry listener
+   */
+  public void addListener(MetricRegisterListener listener) {
+    this.listeners.add(listener);
+  }
+
+  /**
+   * Register a metric
+   *
+   * @param define metric define
+   * @param tags values of tag
+   * @param metric metric
+   */
+  public <T extends Metric> MetricKey register(
+      MetricDefine define, Map<String, String> tags, T metric) {
+    Preconditions.checkNotNull(metric, "Metric must not be null");
+    Preconditions.checkNotNull(define, "Metric define must not be null");
+    Preconditions.checkArgument(
+        define.getType().isType(metric),
+        "Metric type miss-match, required：%s, but found implement:%s ",
+        define.getType(),
+        metric.getClass().getName());
+
+    Pair<MetricDefine, Integer> exists =
+        definedMetric.computeIfAbsent(define.getName(), n -> Pair.of(define, 0));
+    Preconditions.checkArgument(
+        exists.getKey().equals(define),
+        "The metric define with name: %s has been already exists, but the define is different.",
+        define.getName(),
+        exists);
+
+    MetricKey key = new MetricKey(define, tags);
+
+    definedMetric.computeIfPresent(
+        define.getName(),
+        (name, existsDefine) -> {
+          Preconditions.checkArgument(
+              define.equals(existsDefine.getKey()),
+              "Metric define:%s is not equal to existed define:%s",
+              define,
+              existsDefine);
+          Metric existedMetric = registeredMetrics.putIfAbsent(key, metric);
+          Preconditions.checkArgument(existedMetric == null, "Metric is already been registered.");
+          return Pair.of(existsDefine.getKey(), existsDefine.getRight() + 1);
+        });
+
+    callListener(l -> l.onMetricRegistered(key, metric));
+    return key;
+  }
+
+  /**
+   * Remove a metric
+   *
+   * @param key registered metric key
+   */
+  public void unregister(MetricKey key) {
+    Metric exists = registeredMetrics.remove(key);
+    if (exists != null) {
+      callListener(l -> l.onMetricUnregistered(key));
+    }
+    definedMetric.computeIfPresent(
+        key.getDefine().getName(),
+        (n, p) -> {
+          int count = p.getRight() - 1;
+          if (count <= 0) {
+            return null;
+          } else {
+            return Pair.of(p.getKey(), count);
```

Comment:
Maybe we should unify the method calls of the pair class within a class. Now sometimes  using `getRight` and using `getKey`.

Is it more reasonable to use `getKey/getValue` or `getLeft/getRight`?

Reply:
fixed, all use getLeft/getRight